### PR TITLE
udev: add {Receive,Transmit}ChecksumOffload= settings

### LINKED ISSUE #14661: networkd: options for checksum offloading?
**Is your feature request related to a problem? Please describe.**
One of my machines has a buggy network chip or driver, and will stop responding to various types of network traffic after a few hours or days of uptime – unless TX checksum offloading is disabled (`ethtool --offload eth0 tx off`). I'd like to be able to configure this with networkd.

**Describe the solution you'd like**
Addition of a `[Link]` section option for checksum offloading.

**Describe alternatives you've considered**
- Creating a service which runs ethtool to apply the setting (currently using this). It seems like something that networkd should be able to do though.

### PR DESCRIPTION
**Pull request by yuwata**
Closes #14661.

### CODE REVIEW

**Comment by lheckemann** [comment]
I think the names should maybe mention offloading? Since, AFAIU, the checksumming happens regardless, and the setting only determines whether it happens on the CPU or on the NIC.

**Comment by yuwata** [comment]
OK. How about `ReceiveCheckSumOffloading=`?

**Comment by yuwata** [comment]
Renamed. PTAL.

**Comment by flo-wer** [comment]
Following the spelling of other settings, you should name it `ReceiveChecksumOffload=` and `TransmitChecksumOffload=` (cf. https://www.freedesktop.org/software/systemd/man/systemd.directives.html#UDPChecksum=)

**Comment by yuwata** [comment]
@flo-wer Thanks. Renamed again.

**Comment by keszybz** [comment]
LGTM, but let's wait for possible ack from OP.

File: man/systemd.link.xml [hunk 1]
```diff
@@ -612,6 +612,20 @@
           </para>
         </listitem>
       </varlistentry>
+      <varlistentry>
+        <term><varname>ReceiveChecksumOffload=</varname></term>
+        <listitem>
+          <para>Takes a boolean. If set to true, the receive check summing (Rx) is enabled.
```
Comment (poettering) [inline man/systemd.link.xml]: maybe "the receive check summing" → "hardware offload for checksumming of ingress network packets"?

File: man/systemd.link.xml [hunk 2]
```diff
@@ -612,6 +612,20 @@
...
+          When unset, the kernel's default will be used.</para>
+        </listitem>
+      </varlistentry>
+      <varlistentry>
+        <term><varname>TransmitChecksumOffload=</varname></term>
+        <listitem>
+          <para>Takes a boolean. If set to true, the transmit check summing (Tx) is enabled.
```
Comment (poettering) [inline man/systemd.link.xml]: as above

File: src/shared/ethtool-util.c [hunk 3]
```diff
@@ -50,6 +51,8 @@ DEFINE_STRING_TABLE_LOOKUP(port, NetDevPort);
 DEFINE_CONFIG_PARSE_ENUM(config_parse_port, port, NetDevPort, "Failed to parse Port setting");
 
 static const char* const netdev_feature_table[_NET_DEV_FEAT_MAX] = {
+        [NET_DEV_FEAT_RX]   = "rx-checksum",
+        [NET_DEV_FEAT_TX]   = "tx-checksum-*",
```
Comment (poettering) [inline src/shared/ethtool-util.c]: hmm, can you add a comment elaborating on the `*` here?

File: src/shared/ethtool-util.c [hunk 4]
```diff
@@ -498,22 +501,37 @@ static int get_stringset(int ethtool_fd, struct ifreq *ifr, int stringset_id, st
...
-        for (i = 0; i < strings->len; i++) {
-                if (streq((char *) &strings->data[i * ETH_GSTRING_LEN], feature))
-                        return i;
-        }
+        assert(strings);
+        assert(feature);
+        assert(sfeatures);
+
+        for (size_t i = 0; i < strings->len; i++)
+                if (fnmatch(feature, (char *) &strings->data[i * ETH_GSTRING_LEN], 0) == 0) {
```
Comment (poettering) [inline src/shared/ethtool-util.c]: hmm, do we really need globs here? maybe just do a prefix match if feature ends with "-"?

```c
if (streq((char *) &strings->data[i * ETH_GSTRING_LEN], feature) ||
        endswith(feature, "-") && startswith(strings->data[i * ETH_GSTRING_LEN], feature)) …
```

if you get what i mean...

I mean, globbing appears real overkill if all we care about is a prefix match

**Comment by lheckemann** [comment]
Thanks very much for the rapid implementation! I can test this at the weekend.

**Comment by yuwata** [comment]
@poettering Thank you for the suggestions. I've force-pushed the revised version. PTAL.

File: src/shared/ethtool-util.c [hunk 5]
```diff
@@ -1,5 +1,6 @@
 /* SPDX-License-Identifier: LGPL-2.1+ */
 
+#include <fnmatch.h>
```
Comment (poettering) [inline src/shared/ethtool-util.c]: this include can be dropped, now, no?

File: src/shared/ethtool-util.c [hunk 6]
```diff
@@ -50,6 +51,8 @@ DEFINE_STRING_TABLE_LOOKUP(port, NetDevPort);
 DEFINE_CONFIG_PARSE_ENUM(config_parse_port, port, NetDevPort, "Failed to parse Port setting");
 
 static const char* const netdev_feature_table[_NET_DEV_FEAT_MAX] = {
+        [NET_DEV_FEAT_RX]   = "rx-checksum",
+        [NET_DEV_FEAT_TX]   = "tx-checksum-",
```
Comment (poettering) [inline src/shared/ethtool-util.c]: still would like to see a quick comment here that explains the situation, i.e. what the "-" at the end means. Could be as short as `/* The suffix "-" means any feature beginning with "tx-checksum-" */

**Comment by yuwata** [comment]
Addressed two points. Setting the green label.

**Pull request merged**